File: ios/RoutineChart/RoutineChart/Features/Auth/ChildSignInViewModel.swift
//
//  ChildSignInViewModel.swift
//  RoutineChart
//
//  Created for Phase 2.1: Firebase Auth
//

import Foundation
import Combine

@MainActor
final class ChildSignInViewModel: ObservableObject {
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private let authRepository: AuthRepository
    
    init(authRepository: AuthRepository) {
        self.authRepository = authRepository
    }
    
    func signInAsChild() async {
        isLoading = true
        errorMessage = nil
        
        do {
            _ = try await authRepository.signInAnonymously()
            AppLogger.log("Child signed in anonymously")
        } catch {
            AppLogger.error("Child sign in failed", error: error)
            errorMessage = error.localizedDescription
        }
        
        isLoading = false
    }
}

File: android/app/src/main/java/com/HammersTech/RoutineChart/features/auth/child/ChildSignInViewModel.kt
package com.HammersTech.RoutineChart.features.auth.child

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.HammersTech.RoutineChart.core.domain.repositories.AuthRepository
import com.HammersTech.RoutineChart.core.utils.AppLogger
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel for child sign-in
 * Phase 2.1: Firebase Auth
 */
@HiltViewModel
class ChildSignInViewModel
    @Inject
    constructor(
        private val authRepository: AuthRepository,
    ) : ViewModel() {
        private val _state = MutableStateFlow(ChildSignInState())
        val state: StateFlow<ChildSignInState> = _state.asStateFlow()

        fun signInAsChild() {
            viewModelScope.launch {
                _state.update { it.copy(isLoading = true, errorMessage = null) }

                val result = authRepository.signInAnonymously()

                result.fold(
                    onSuccess = {
                        AppLogger.UI.info("Child signed in anonymously")
                        _state.update { it.copy(isLoading = false) }
                    },
                    onFailure = { error ->
                        AppLogger.UI.error("Child sign in failed", error)
                        _state.update {
                            it.copy(
                                isLoading = false,
                                errorMessage = error.message ?: "Sign in failed",
                            )
                        }
                    },
                )
            }
        }
    }

data class ChildSignInState(
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
)
